pkg/scheduler/framework/cycle_state.go

```diff
@@ -0,0 +1,85 @@
+package framework
+
+import (
+	"errors"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+var (
+	// ErrNotFound is the not found error message.
+	ErrNotFound = errors.New("not found")
+)
+
+// StateData is a generic type for arbitrary data stored in CycleState.
+type StateData interface {
+	// Clone is an interface to make a copy of StateData. For performance reasons,
+	// clone should make shallow copies for members (e.g., slices or maps) that are not
+	Clone() StateData
+}
+
+// StateKey is the type of keys stored in CycleState.
+type StateKey string
+
+// CycleState provides a mechanism for plugins to store and retrieve arbitrary data.
+// StateData stored by one plugin can be read, altered, or deleted by another plugin.
+// CycleState does not provide any data protection, as all plugins are assumed to be
+// trusted.
+// Note: CycleState uses a sync.Map to back the storage. It's aimed to optimize for the "write once and read many times" scenarios.
+// It is the recommended pattern used in all in-tree plugins - plugin-specific state is written once in PreFilter/PreScore and afterwards read many times in Filter/Score.
+type CycleState struct {
+	// storage is keyed with StateKey, and valued with StateData.
+	storage sync.Map
```

Comment:
Sorry, I can't figure out how to use this filed, can you help explain how to use this field?

Reply:
Like: 

```
type Clusters []*clusterv1alpha1.Cluster

// Clone the Clusters state.
func (s Clusters) Clone() framework.StateData {
	return s
}


cycleState.Write("clustersinfo", Clusters(xxxx))

cycleState.Read("clustersinfo")


```

Reply:
In the future, I will store AvailableReplicas in CycleState using this method.

Reply:
Does this mean we need to add some pre-plug-in?

Reply:
Yes. 

Refer to https://github.com/karmada-io/karmada/pull/3287#issuecomment-1477198987

Reply:
We plan to introduce a pre-plugin and a score plugin in the future.

Reply:
Should we design and analyze the introduction of the pre-plugin and the addition of the current `CycleState` structure as a whole, so that we can see the whole landscape of the optimization process, rather than the current single-point-triggered improvement?

Reply:
You can see that under the current scheduling framework, it is so difficult to add a new plugin. As @Garrybest  said, it is time to optimize our scheduling framework.

Reply:
I understand that. I mean, shouldn't we come up with a design plan, then make a plan, split the tasks and implement them one by one?

Reply:
> Should we design and analyze the introduction of the pre-plugin and the addition of the current CycleState structure as a whole

It is necessary to introduce CycleState, regardless of whether pre-filter or pre-score or other plug-ins are added.

Reply:
> I mean, shouldn't we come up with a design plan, then make a plan, split the tasks and implement them one by one?

+1

Reply:
Exactly. We need a task todo list. It is a big job for us to make an optimization for scheduler framework.